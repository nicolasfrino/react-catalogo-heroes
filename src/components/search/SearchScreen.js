import React, { useMemo } from 'react'
import { useLocation } from 'react-router-dom';
//import { heroes } from '../../data/heroes'
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../hooks/useForm';
import queryString from 'query-string';
import { getHeroesByName } from '../../selectors/getHeroesByName';



export const SearchScreen = ( { history} ) => {

    const location = useLocation();
    //console.log('location', location);
    const {q = ''} = queryString.parse(location.search);
    //console.log('q', q);



    const [formValues, handleInputChange] = useForm({
        searchText: q
    })

    const { searchText } = formValues;

    //    const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher ]);
    const heroFiltered = useMemo(() => getHeroesByName(q), [q]);

/* aca el curso me pude que use el memo, no se puede usar el efec?
    useEffect(() => {
        //console.log('searchText cambio');
        return () => {
        }
    }, [searchText]);
*/

    const handleSearch = (e) => {
        e.preventDefault();

        //console.log(history);
        history.push(`?q=${ searchText }`)
    }

    return (
        <div>
            <h1>SearchScreen</h1>
            <hr/>

            <div className="row">

                <div className="col-5">
                    <h4>SearchForm</h4>
                    <form onSubmit={handleSearch}>
                        <input  type="text"
                                name="searchText"
                                autoComplete="off"
                                value= {searchText}
                                placeholder="Busca tu Heroe"
                                className="find-control"
                                onChange={handleInputChange}
                        />
                        <button type="submit"
                                className="btn m-1 btn-block btn-outline-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>
                    { (q === '')
                        &&
                        <div className="alert alert-info">
                            Buscar un Heroe
                        </div>
                    }
                    { (q !== '' && heroFiltered.length === 0)
                        &&
                        <div className="alert alert-danger">
                            {`no hay heroe con valores: ${ q }`}
                        </div>
                    }
                    {
                        heroFiltered.map( heroRet => ( <HeroCard key={heroRet.id} {...heroRet}/> ))
                    }
                </div>

            </div>

        </div>

    )
}
