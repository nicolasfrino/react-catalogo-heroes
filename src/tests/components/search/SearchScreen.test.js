import { mount } from "enzyme";
import React from "react";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe('Pruebas en Search', () => {
    
    test('Mostrar valores por defecto', () => {
        
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search']}>
                <Route exact path="/search" component= { SearchScreen} />
            </MemoryRouter>
        );

        expect (wrapper).toMatchSnapshot();
        expect (wrapper.find('.alert-info').text().trim()).toBe('Buscar un Heroe');

    })
    
    test('Mostrar a batman y el input con el valor de qstring', () => {
        
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route exact path="/search" component= { SearchScreen} />
            </MemoryRouter>
        );

        expect (wrapper.find('input').prop('value')).toBe('batman');
        expect (wrapper).toMatchSnapshot();


    })

    test('Mostar error si no es encuentra el hero', () => {
        
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route exact path="/search" component= { SearchScreen} />
            </MemoryRouter>
        );

        expect (wrapper.find('input').prop('value')).toBe('batman123');
        expect (wrapper.find('.alert-danger').text().trim()).toBe('no hay heroe con valores: batman123');
        

    })

    test('debe llamar al push del history', () => {
        
        const historyMock = {
            push: jest.fn()
        }

        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    exact path="/search" 
                    component= { () => <SearchScreen history={historyMock}/> } 

                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change',{
            target: {
                name: 'searchText',
                value: 'batman'
                }
            }
        );

        wrapper.find('form').simulate('submit',{
                preventDefault() {}
            }
        );

        //wrapper.find('form').prop('onSubmit')( {
        //    preventDefault() {}
        //});

        expect(historyMock.push).toHaveBeenCalledWith('?q=batman')

    })

})
