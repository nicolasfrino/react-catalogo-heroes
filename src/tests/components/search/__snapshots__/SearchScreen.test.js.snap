// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Search Mostrar a batman y el input con el valor de qstring 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        SearchForm
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="find-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Busca tu Heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src=" ./assets/heroes/dc-batman.jpg "
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                 
                Batman
                 
              </h5>
              <p
                className="card-text"
              >
                   
                Bruce Wayne
                 
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                   
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en Search Mostrar valores por defecto 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        SearchForm
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="find-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Busca tu Heroe"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un Heroe
      </div>
    </div>
  </div>
</div>
`;
